### PR TITLE
[@types/node] SpawnSyncReturns may have null values

### PR DESCRIPTION
**Pull request by KSXGitHub**
### When spawn a program that doesn't exist

```javascript
const res = child_process.spawnSync('UnknownProgram')
res.status // null
res.signal // null
res.output // null
res.stdout // null
res.stderr // null
res.Error // Error
```

### When successfully spawn a program

```javascript
const res = child_process.spawnSync('echo', ['hello'])
res.Error // null
res.signal // null
res.output // [null, Buffer('hello\n'), Buffer(0)]
```

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by KSXGitHub** [comment]
Although CI had failed at [`cross-spawn`](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/cross-spawn), I'm sure this change is correct, some one should update `cross-spawn` accordingly.

**Comment by KSXGitHub** [comment]
@Alorel please review https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25731/commits/61cbba608786a83523e879972b966c3134b235b6

**Review by Flarna — APPROVED** [review]

**Comment by KSXGitHub** [comment]
Breaking change is expected (I had to also edit `@types/cross-spawn` to let it pass). So you might want to assign an appropriate version.

**Comment by Flarna** [comment]
@KSXGitHub There is no semver versioning on typings... see #25677 for details.
So your change will go into next patch version of node 8, 9 and 10. If this is not intended the PR needs to be adapted.

**Comment by KSXGitHub** [comment]
Personally, I think this PR is necessary. The very-purpose purpose of TypeScript is type-safety, the old definition fails to provided safety by wrongly assume `SpawnSyncReturns` always not-null which is not the case. Therefore I think this PR should be merged. If it breaks other people builds, so be it, they just get a warning of a potential bug that might occur in the future.

**Comment by Flarna** [comment]
I'm not 100% sure here.
It's similar for common callback like APIs, e.g. `fs.readFile()` which has currently following callback signature `(err: NodeJS.ErrnoException, data: Buffer) => void).
Actually correct would be `(err: NodeJS.ErrnoException | null , data: Buffer | undefined) => void` as `err` may be `null` (in case of success) and `data` may be undefined (in case of failure).

Therefore if we would change this and you check in your code that `err === null` would be additionally needed to check that `data` is not `undefined`.

**Comment by KSXGitHub** [comment]
I made this pull request because I encountered a very bullshit bug in my project due to the use of `spawnSync`. I am sure that the old definition is nowhere near usability.

File: types/node/index.d.ts [hunk 1]
```diff
@@ -2333,12 +2333,12 @@ declare module "child_process" {
...
-        stderr: T;
-        status: number;
-        signal: string;
-        error: Error;
+        output: [T | null, T | null, T | null] | null;
+        stdout: T | null;
+        stderr: T | null;
+        status: number | null;
+        signal: string | null;
+        error: Error | null;
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply types/node/index.d.ts]: Maybe, I'm 100% sure how exactly `spawnSync` works. But I am sure though, when `error` is not null, the other properties are. I'm not so sure about `signal`.

[the remark this answers is not in the document]
Comment (KSXGitHub) [reply types/node/index.d.ts]: I also hope that no fools create an `interface` that extends this one. :smile:

[the remark this answers is not in the document]
Comment (KSXGitHub) [reply types/node/index.d.ts]: @ajafff Your thought on [this](https://github.com/KSXGitHub/DefinitelyTyped/blob/aa2c5741f98ae8cee08b433d6df25aa11bb82bb9/types/node/index.d.ts#L2334-L2355) ([diff](https://github.com/KSXGitHub/DefinitelyTyped/commit/58aed8ac63d20490207520d3d3f5494e683c53da))

File: types/node/index.d.ts [hunk 2]
```diff
@@ -2331,14 +2331,27 @@ declare module "child_process" {
...
     }
-    export interface SpawnSyncReturns<T> {
-        pid: number;
-        output: string[];
-        stdout: T;
-        stderr: T;
-        status: number;
-        signal: string;
-        error: Error;
+    export type SpawnSyncOptions = SpawnSyncReturns.WithError | SpawnSyncOptions.WithoutError<T>;
```
Comment (ajafff) [inline types/node/index.d.ts]: this should be `SpawnSyncReturns` consistently

**Review by ajafff — CHANGES_REQUESTED** [review]

File: types/node/index.d.ts (hunk 2 above)
Comment (KSXGitHub) [reply types/node/index.d.ts]: Oh, my bad, I chose wrong options in my autocomplete menu. I'll fix this ASAP.

Comment (KSXGitHub) [reply types/node/index.d.ts]: Done.

Comment (KSXGitHub) [reply types/node/index.d.ts]: Damn, I always make mistakes.

**Comment by KSXGitHub** [comment]
@Flarna This might be off-topic. But since you mentioned callback types of `fs.readFile`, I would like to give my opinion. Yeah, the correct parameter type should have had `null`, one of the reason that `error` is the first parameter is because error **should not** be ignore silently (I read it somewhere when I learn Node.js), so giving `null` to `error` is the right choice. As for `data`, it may not have `null` because user should always look for `error` instead. Therefore the correct type should be `(error: NodeJS.ErrnoException | null, data: Buffer) => void`.

Thankfully, there's an alternate solution to read file asynchronously: promises and async function. With it, I can treat `NodeJS.ErrnoException` the same way I treat normal JavaScript errors.

File: types/cross-spawn/cross-spawn-tests.ts [hunk 3]
```diff
@@ -8,13 +8,13 @@ pw.on('error', (err: Error) => {
...
-const r5: number = spawn.sync('foo').status;
-const r6: string = spawn.sync('foo').signal;
-const r7: string = spawn.sync('foo', {encoding: 'utf8'}).stdout;
+const r1: Error | null = spawn.sync('foo').error;
+const r2: [Buffer | null, Buffer | null, Buffer | null] | null = spawn.sync('foo').output;
+const r3: Buffer | null = spawn.sync('foo').stdout;
+const r4: Buffer | null = spawn.sync('foo').stderr;
+const r5: number | null = spawn.sync('foo').status;
+const r6: string | null = spawn.sync('foo').signal;
+const r7: string | null = spawn.sync('foo', {encoding: 'utf8'}).stdout;
```
Comment (ajafff) [inline types/cross-spawn/cross-spawn-tests.ts]: can you also add a test to ensure the type is correctly inferred after testing if `error == null`?

**Review by ajafff — APPROVED** [review]

File: types/cross-spawn/cross-spawn-tests.ts (hunk 3 above)
Comment (KSXGitHub) [reply types/cross-spawn/cross-spawn-tests.ts]: Done.

**Comment by KSXGitHub** [comment]
There's still a thing though. If I set option `stdio` to `'inherit'`, all properties except `status` and `pid` will be `null`.

**Comment by KSXGitHub** [comment]
And as I tested in my machine, if `signal` is `SIGKILL`, all properties except `signal`, `pid` and `output` will be `null`, `output` is `[null, null, null]`.

It was in my machine (Linux), can anyone test it in other machine?

**Comment by KSXGitHub** [comment]
@ajafff So, about updating `cross-spawn-tests.ts`, we got a [problem](https://www.typescriptlang.org/play/#src=type%20ValueCarrier%3CX%3E%20%3D%0A%20%20%20%20ValueCarrier.WithoutError%3CX%3E%20%7C%20ValueCarrier.WithError%0A%0Anamespace%20ValueCarrier%20%7B%0A%20%20%20%20export%20interface%20WithError%20%7B%0A%20%20%20%20%20%20%20%20error%3A%20Error%0A%20%20%20%20%20%20%20%20value%3A%20null%0A%20%20%20%20%7D%0A%0A%20%20%20%20export%20interface%20WithoutError%3CValue%3E%20%7B%0A%20%20%20%20%20%20%20%20error%3A%20null%0A%20%20%20%20%20%20%20%20value%3A%20Value%0A%20%20%20%20%7D%0A%7D%0A%0Adeclare%20function%20getVal%3CX%3E()%3A%20ValueCarrier%3CX%3E%0A%0A%2F*%20MAIN%20*%2F%0Aconst%20carrier%20%3D%20getVal%3Cstring%3E()%0A%0Aif%20(carrier.error%20%3D%3D%20null)%20%7B%0A%20%20%20%20console.error(carrier.error)%0A%20%20%20%20const%20value%3A%20null%20%3D%20carrier.value%20%2F%2F%20Error%20here...%0A%7D%20else%20%7B%0A%20%20%20%20const%20value%3A%20string%20%3D%20carrier.value%0A%20%20%20%20console.info('Value'%2C%20value)%0A%7D%0A).

```typescript
type ValueCarrier<X> =
    ValueCarrier.WithoutError<X> | ValueCarrier.WithError

namespace ValueCarrier {
    export interface WithError {
        error: Error
        value: null
    }

    export interface WithoutError<Value> {
        error: null
        value: Value
    }
}

declare function getVal<X>(): ValueCarrier<X>

/* MAIN */
const carrier = getVal<string>()

if (carrier.error == null) {
    console.error(carrier.error)
    const value: null = carrier.value // Error here...
} else {
    const value: string = carrier.value
    console.info('Value', value)
}
```

Perhaps I should restore the old version.

**Comment by KSXGitHub** [comment]
@ajafff Since `Error | null` is somehow equivalent to just `Error`, tagged union is pretty useless.

**Comment by ajafff** [comment]
@KSXGitHub I agree. If TypeScript doesn't handle it correctly, using tagged unions is unnecessary and should be reverted.

**Comment by KSXGitHub** [comment]
@ajafff Since reverting it make practically no differences, I would like to keep it in hope that TypeScript will fix it soon.

**Comment by ajafff** [comment]
> `"strictNullChecks": false`

I don't know why. But I do know that this causes the tagged union to not work as expected.

**Comment by Flarna** [comment]
Because node type definitions have been created before typescript 2.0 (and therefore strictNullChecks) was released and till now nobody did the work to fix all the issues which pop up if it is set to true. I can remember that someone started a while ago but didn't complete.
Similar the corrections of the typings for callbacks to e.g. `Error | null` is also to be done and has the same root cause.

**Comment by DanielRosenwasser** [comment]
So just to summarize for myself, this is a breaking change and currently TypeScript doesn't use `null` as a discriminator, so you can't appropriately check the contents. Is that correct?

**Comment by KSXGitHub** [comment]
@DanielRosenwasser Yes.

**Comment by KSXGitHub** [comment]
@Flarna 

> So your change will go into next patch version of node 8, 9 and 10.

Node.js v10.2.0 just released.

**Comment by Flarna** [comment]
@KSXGitHub The release cycle of NodeJS and of @types/node is not related. NodeJS exposes/documents javascript APIs and doesn't care about typescript.
Whenever a PR here gets merged the types published creates a new @types/XXX release using the major.minor from the index.d.ts file and increments patch. This may happen several times per Node release of there may happen several nodejs releases without any updates in type definitions.

**Pull request closed without merging**